### ISSUE TITLE
The filter seems to corrupt certain files

### ISSUE THREAD

**Issue by julienfromentc** [rgielen/struts1filter#2]
Hello,

After testing the latest filter, we found some case where when we upload certain files, the files that gets saved is "corrupted".

You can find an offending file here: https://github.com/julienfromentc/struts1filter-issue/raw/master/PDF.PDF, in this case the resulting PDF will be blank page.

**Comment by pwntester** [comment]
Sorry for the delay.

I tried settting the request encoding to UTF-8 before reading the inputStream: request.setCharacterEncoding("utf-8");

I also added the original encoding to the InputStreamReader:
String characterEncoding = this.getCharacterEncoding();
                    if (characterEncoding == null ) {
                        bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
                    }
                    else {
                        bufferedReader = new BufferedReader(new InputStreamReader(inputStream, characterEncoding));
                    }

But I still get some bytes converted. Note that those bytes are read that way by inputStream. You can get the bytes from body and print them and you will see that the bytes after %PDF-1.2% are efbfbdef ... instead of e2e3cfe30d0a...

@rgielen any ideas?

**Comment by julienfromentc** [comment]
@pwntester @rgielen Did you guys had any chances to look at this issue a bit?

**Comment by pwntester** [comment]
nothing new since last update. @rgielen can you please look at it? Any idea if Struts is finally going to release a new version using the secure BeansUtil version??

**Comment by rajeshmanoharann** [comment]
In below PR, which is raised to struts repo, has change in RequestUtils which checks request.getParameterNames alone. But this filter checks input stream also. Is it really needed to check input stream?

https://github.com/apache/struts1/pull/1

**Comment by pwntester** [comment]
Yep, it is since payload can be placed in a multipart post request

**Comment by pwntester** [comment]
I think using a fixed version of Struts 1 is a much better option than using the filter since unfortunately its not as stable as we would like it to be. And actually applying a patch should not be that difficult. The official patch uses a new version of BeanUtils that basically applies the regular expression. It should be similar to:
https://gist.github.com/ykare/1fb2d5d11e16cb0003c3#file-2-beanutils-patch-struts1-0-2

But instead of patching BeanUtils (used by Struts), you can directly patch Struts library, like this pull request proposes:
https://github.com/apache/struts1/pull/1/files

Another alternative is to download a struts version that comes with the patch. Fedora has applied it:
http://koji.fedoraproject.org/koji/buildinfo?buildID=552377

**Comment by jimnz111** [comment]
I'm quite late in chiming in on this issue. We had to implement the filter temporarily to pass security requirements. I too came across this issue for multipart form data. Tracked issue back to the conversion of the InputStream which is byte[] array to a char[] array. You cannot convert binary data to char and back again and hope it won't change! The fix for this is to instead read the full byte[] stream and store that, whilst still doing conversion of body to char array to check for malicious strings. The subsequent call in ``getInputStream()`` should return the original byte[] array instead of body.

Side note that the 1.0.1-SNAPSHOT also doesn't work. It appears that the removal of the call to super.getParameterNames() in constructor causes the filter to fail! I haven't debugged through to see why its needed, however leaving it in keeps everything working

**Comment by ilatypov** [comment]
Thank you @jimnz111 for noticing the input swallowing in the 1.0.1-SNAPSHOT filter and analyzing the cause.

Using Alberto Fernández's maintenance branch of Struts 1 which included [the fix in ActionServlet](https://github.com/albfernandez/struts1/commit/c7f3209e9241df397f46b2d6ba0464969106d75c) and [the fix in RequestUtils](https://github.com/albfernandez/struts1/commit/5a48a99aad44f7c865065a30449f68467535d643) protected against the RCE ".class." attack.

The mass assignment of business-sensitive fields (IDOR) needs addressing by a [white-listing technique](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa6-mass-assignment.md#how-to-prevent) such as defining dedicated Data Transfer Objects.

P.S. A hack in doFilter (after undoing the Jan 22, 2015 change) can preserve binary uploads at the expense of corrupting non-UTF8 text inputs.
```java
        request.setCharacterEncoding("UTF-8");
        response.setCharacterEncoding("UTF-8");
```
